xmppserver/src/main/java/org/jivesoftware/openfire/pubsub/CachingPubsubPersistenceProvider.java

```diff
@@ -0,0 +1,708 @@
+    {
+        delegate.createDefaultConfiguration( service, config );
+    }
+
+    @Override
+    public void updateDefaultConfiguration( final PubSubService service, final DefaultNodeConfiguration config )
+    {
+        delegate.updateDefaultConfiguration( service, config );
+    }
+
+    @Override
+    public List<PublishedItem> getPublishedItems( final LeafNode node )
+    {
+        flushPendingItems( node.getUniqueIdentifier() );
+        return delegate.getPublishedItems( node );
+    }
+
+    @Override
+    public List<PublishedItem> getPublishedItems( final LeafNode node, final int maxRows )
+    {
+        flushPendingItems( node.getUniqueIdentifier() );
+        return delegate.getPublishedItems( node, maxRows );
+    }
+
+    @Override
+    public PublishedItem getLastPublishedItem( final LeafNode node )
+    {
+        flushPendingItems( node.getUniqueIdentifier() );
+        return delegate.getLastPublishedItem( node );
+    }
+
+    @Override
+    public PublishedItem getPublishedItem( final LeafNode node, final PublishedItem.UniqueIdentifier itemIdentifier )
+    {
+        flushPendingItems( node.getUniqueIdentifier() );
+
+        // try to fetch from cache first without locking
+        PublishedItem result = itemCache.get(itemIdentifier);
+        if (result == null) {
+            Lock itemLock = CacheFactory.getLock( ITEM_CACHE, itemCache);
+            try {
+                // Acquire lock, then re-check cache before reading from DB;
+                // allows clustered item cache to be primed by first request
+                itemLock.lock();
+                result = itemCache.get(itemIdentifier);
+                if (result == null) {
+                    log.debug("No cached item found. Obtaining it from delegate.");
+                    result = delegate.getPublishedItem( node, itemIdentifier );
+                    itemCache.put(itemIdentifier, result);
+                } else {
+                    log.debug("Found cached item on second attempt (after acquiring lock)");
+                }
+
+            } finally {
+                itemLock.unlock();
+            }
+        } else {
+            log.debug("Found cached item on first attempt (no lock)");
+        }
+        return result;
+    }
+
+    static class NodeOperation {
+
+        enum Action {
+            CREATE,
+            UPDATE,
+            REMOVE,
+            CREATE_AFFILIATION,
+            UPDATE_AFFILIATION,
+            REMOVE_AFFILIATION,
+            CREATE_SUBSCRIPTION,
+            UPDATE_SUBSCRIPTION,
+            REMOVE_SUBSCRIPTION
+        }
+
+        final Node node;
+        final Action action;
+        final NodeAffiliate affiliate;
+        final NodeSubscription subscription;
+
+        private NodeOperation( final Node node, final Action action, final NodeAffiliate affiliate, final NodeSubscription subscription ) {
+            if ( node == null ) {
```

Comment:
`this.node = Objects.requireNonNull(node, "Argument 'node' cannot be null.");`
replaces four lines with one.

Reply:
Meh. I've been using this consistently. We might at one point want to consistently do a search/replace and replace the lot, but I'll leave things as-they-are for now.

Related: what are your thoughts on using `@javax.annotation.Nonnull` and `@javax.annotation.Nullable`? Intellij seems to work with them in a nice way.

Reply:
Yes! I use Nonnull / Nullable liberally throught my code, combined with Optional<> return values to help cut down on my mistakes.